[SPEC] Inverse operation (#22786)
openvinotoolkit/openvino: docs/articles_en/documentation/openvino-ir-format/operation-sets/operations-specifications/matrix/Inverse_14.rst

.. {#openvino_docs_ops_matrix_Inverse_14}

Inverse
=======

.. meta::

>   :description: Learn about Inverse-14 - a matrix operation that computes the inverse or adjoint of one or multiple input matrices.

Reviewer: ```suggestion
  :description: Learn about Inverse-14 - a matrix operation that computes the inverse or adjoint for a batch of input matrices.
```

Author: Applied!

**Versioned name**: *Inverse-14*

**Category**: *Matrix*

> **Short description**: *Inverse* operation computes either the inverse or adjoint (conjugate transposes) of one or more square invertible matrices.

Reviewer: also better to say about batch of matrices

Author: Added!

> **Detailed description**: *Inverse* operation computes the inverse of a square matrix. The operation uses LU decomposition with partial pivoting to compute the inverses.

Reviewer: Can we describe here a reference algorithm for LU decomposition? Or at least leave a link to it?

Author: For now I have added the link to the algorithm description, I will add the full algorithm pseudo code after CF!

The inverse matrix A^(-1) of a square matrix A is defined as:

.. math::

   A \cdot A^{-1} = A^{-1} \cdot A = I

where **I** is the n-dimensional identity matrix.

The inverse matrix exists if and only if the input matrix is invertible. In that case, the inverse is unique. If the matrix is not invertible however, the operation may raise an exception or return an undefined result.

This operation can be used to compute the adjugate matrix instead of the inverse.

The adjugate matrix adj(A) of a square matrix A is defined as:

.. math::

   adj(A) = det(A) \cdot A^{-1}

where **A^{-1}** is the matrix inverse of A, and **det(A)** is the determinant of A.

The adjugate matrix exists if and only if the inverse matrix exists.

> **Inputs**:

Reviewer: since only one input
```suggestion
**Input**:
```

Author: Applied!

> * **1**: `input` - A tensor of shape [B1, B2, ..., Bn, ROW, COL] and type `T` representing the input square matrices. The inner-most 2 dimensions form square matrices and must be of the same size. **Required.**

Reviewer: `B1, B2, ..., Bn` are the batch dimensions

Author: Described!

> **Outputs**:

Reviewer: ```suggestion
**Output**:
```

Author: Applied!

* **1**: `output` - A tensor with the same type `T` as the input and same shape [B1, B2, ..., Bn, ROW, COL] as the input, representing the inverse matrices (or adjugate matrices) of the input matrices.

> **Attributes**:

Reviewer: ```suggestion
**Attribute**:
```

Author: Applied!

* *adjoint*

  * **Description**: Modifies the return value of the operation. If true, the operation returns the adjoint (conjugate transpose) of the input matrices instead of finding the inverse.
  * **Range of values**: `true`, `false` 

    * ``true`` - output adjugate matrix.
    * ``false`` - output inverse matrix. 

  * **Type**: `bool`
  * **Default value**: `false`

>   * **Required**: *No*

Reviewer: please put attribute section above inputs as we do for other ops

Author: Applied!

**Types**

* **T**: any supported floating-point type.

*Example 1: 2D input matrix.*

.. code-block:: xml
    :force:

    <layer ... name="Inverse" type="Inverse">
        <data/>
        <input>
            <port id="0" precision="FP32">
                <dim>3</dim> <!-- matrix size of 3x3 -->
                <dim>3</dim> 
            </port>
        </input>
        <output>
            <port id="1" precision="FP32" names="Inverse:0">

>                 <dim>3</dim> <!-- matrix size of 3x3 -->
>                 <dim>3</dim> 

Reviewer 2: As the two last dimensions are specified as [ROW, COL],
I would suggest to reflect that in the examples like:
```suggestion
                <dim>3</dim> <!-- rows of square matrix -->
                <dim>3</dim> <!-- columns of square matrix -->
```

Author: Applied!
